File: model-client/src/main/java/org/modelix/model/lazy/CLHamtLeaf.java
package org.modelix.model.lazy;

import org.apache.commons.lang3.mutable.MutableObject;
import org.modelix.model.persistent.CPHamtLeaf;
import org.modelix.model.persistent.CPHamtNode;
import org.modelix.model.persistent.HashUtil;

import java.util.Objects;
import java.util.function.BiPredicate;

public class CLHamtLeaf extends CLHamtNode<CPHamtLeaf> {

  public static CLHamtLeaf create(long key, String value, IDeserializingKeyValueStore store) {
    return (value == null ? null : new CLHamtLeaf(key, value, store));
  }

  private final CPHamtLeaf data;

  public CLHamtLeaf(CPHamtLeaf data, IDeserializingKeyValueStore store) {
    super(store);
    this.data = data;
  }

  private CLHamtLeaf(long key, String value, IDeserializingKeyValueStore store) {
    super(store);
    this.data = new CPHamtLeaf(key, value);
    String serialized = this.data.serialize();
    store.put(HashUtil.sha256(serialized), data, serialized);
  }

  @Override
  public CPHamtNode getData() {
    return data;
  }

  public long getKey() {
    return data.getKey();
  }

  public String getValue() {
    return data.getValue();
  }

  @Override
  public CLHamtNode put(long key, String value, int shift) {
    if (Objects.equals(key, data.key)) {
      if (Objects.equals(value, data.value)) {
        return this;
      } else {
        return CLHamtLeaf.create(key, value, store);
      }
    } else {
      if (shift > CLHamtNode.MAX_SHIFT) {
        throw new RuntimeException(shift + " > " + CLHamtNode.MAX_SHIFT);
      }
      CLHamtNode result = createEmptyNode();
      result = result.put(data.key, data.value, shift);
      if (result == null) {
        result = createEmptyNode();
      }
      result = result.put(key, value, shift);
      return result;
    }
  }

  @Override
  public CLHamtNode remove(long key, int shift) {
    if (Objects.equals(key, data.key)) {
      return null;
    } else {
      return this;
    }
  }

  @Override
  public IBulkQuery.Value<String> get(long key, int shift, IBulkQuery bulkQuery) {
    return bulkQuery.constant((Objects.equals(data.key, key) ? data.value : null));
  }

  @Override
  public boolean visitEntries(BiPredicate<Long, String> visitor) {
    return visitor.test(data.key, data.value);
  }

  @Override
  public void visitChanges(CLHamtNode oldNode, final CLHamtNode.IChangeVisitor visitor) {
    if (oldNode == this) {
      return;
    }

    final MutableObject<String> oldValue = new MutableObject<>();
    oldNode.visitEntries(new BiPredicate<Long, String>() {
      public boolean test(Long k, String v) {
        if (Objects.equals(k, data.key)) {
          oldValue.setValue(v);
        } else {
          visitor.entryRemoved(k, v);
        }
        return true;
      }
    });

    if (oldValue.getValue() == null) {
      visitor.entryAdded(data.key, data.value);
    } else if (oldValue.getValue() != data.value) {
      visitor.entryChanged(data.key, oldValue.getValue(), data.value);
    }
  }
}

File: model-client/src/main/kotlin/org/modelix/model/lazy/CLHamtLeaf.kt
package org.modelix.model.lazy

import org.apache.commons.lang3.mutable.MutableObject
import org.modelix.model.persistent.CPHamtLeaf
import org.modelix.model.persistent.CPHamtNode
import org.modelix.model.persistent.HashUtil
import java.util.function.BiPredicate

class CLHamtLeaf : CLHamtNode<CPHamtLeaf?> {
    private val data: CPHamtLeaf

    constructor(data: CPHamtLeaf, store: IDeserializingKeyValueStore?) : super(store!!) {
        this.data = data
    }

    private constructor(key: Long, value: String, store: IDeserializingKeyValueStore) : super(store) {
        data = CPHamtLeaf(key, value)
        val serialized = data.serialize()
        store.put(HashUtil.sha256(serialized), data, serialized)
    }

    override fun getData(): CPHamtNode {
        return data
    }

    val key: Long
        get() = data.key

    val value: String
        get() = data.value

    override fun put(key: Long, value: String?, shift: Int): CLHamtNode<*>? {
        return if (key == data.key) {
            if (value == data.value) {
                this
            } else {
                create(key, value, store)
            }
        } else {
            if (shift > MAX_SHIFT) {
                throw RuntimeException("$shift > $MAX_SHIFT")
            }
            var result: CLHamtNode<*>? = createEmptyNode()
            result = result!!.put(data.key, data.value, shift)
            if (result == null) {
                result = createEmptyNode()
            }
            result = result.put(key, value, shift)
            result
        }
    }

    override fun remove(key: Long, shift: Int): CLHamtNode<*>? {
        return if (key == data.key) {
            null
        } else {
            this
        }
    }

    override fun get(key: Long, shift: Int, bulkQuery: IBulkQuery?): IBulkQuery.Value<String?> {
        return bulkQuery!!.constant<String?>(if (data.key == key) data.value else null)!!
    }

    override fun visitEntries(visitor: BiPredicate<Long?, String?>?): Boolean {
        return visitor!!.test(data.key, data.value)
    }

    override fun visitChanges(oldNode: CLHamtNode<*>?, visitor: IChangeVisitor?) {
        if (oldNode === this) {
            return
        }
        val oldValue = MutableObject<String?>()
        val bp: BiPredicate<Long?, String?> = BiPredicate { k, v ->
            if (k == data.key) {
                oldValue.setValue(v)
            } else {
                visitor!!.entryRemoved(k!!, v)
            }
            true
        }
        oldNode!!.visitEntries(bp)
        if (oldValue.value == null) {
            visitor!!.entryAdded(data.key, data.value)
        } else if (oldValue.value !== data.value) {
            visitor!!.entryChanged(data.key, oldValue.value, data.value)
        }
    }

    companion object {
        fun create(key: Long, value: String?, store: IDeserializingKeyValueStore): CLHamtLeaf? {
            return value?.let { CLHamtLeaf(key, it, store) }
        }
    }
}
